File: assign1/src/Main.java
import java.util.Scanner;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;


public class Main {

    public static void printToCSV(List<Double> times, String fileName) {

        try {
            PrintWriter writer = new PrintWriter(fileName);
            double average = 0;
            StringBuilder sb = new StringBuilder();

            // Average
            for (Double item : times) average += item;
            average /= times.size();

            sb.append("Iteration;Execution Time\n");
            writer.write(sb.toString());

            sb.setLength(0);
            for (int iteration = 0; iteration < times.size(); iteration++) {
                sb.append(iteration);
                sb.append(';');
                sb.append(times.get(iteration));
                sb.append('\n');
                writer.write(sb.toString());
                sb.setLength(0);
            }

            sb.append("Average;");
            sb.append(average);
            writer.write(sb.toString());

            writer.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public static void partiallyPrintMatrix(int[][] matrix, int size) {
        int val = 6;
        if (size < val) val = size;

        System.out.println("\n\nMatrix (partially):");

        for (int i = 0; i < val; i++) {
            System.out.print('|');
            for (int j = 0; j < val; j++) {
                System.out.print(' ');
                System.out.print(matrix[i][j]);
                System.out.print(' ');
            }
            System.out.println('|');
        }
    }

    public static void initializeMatrixes(int[][] m1, int[][] m2, int[][] res, int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                m1[i][j] = 1;
                m2[i][j] = i + 1;
                res[i][j] = 0;
            }
        }
    }

    public static void benchmark() {


        for (int size = 600; size <= 3000; size+=400) {      
            int[][] m1 = new int[size][size];
            int[][] m2 = new int[size][size];
            int[][] res = new int[size][size];
            Main.initializeMatrixes(m1, m2, res, size);

            // Mult
            List<Double> results = new ArrayList<>();
            for (int iteration = 0; iteration < 3; iteration++) {
                double start = System.nanoTime();
                Main.OnMult(m1, m2, res, size);
                double stop = System.nanoTime();
                results.add((stop - start) / 1000000000);
            }
            StringBuilder sb = new StringBuilder();
            sb.append("../doc/execution_data/mult/java_mult_result_");
            sb.append(size);
            sb.append(".csv");
            Main.printToCSV(results, sb.toString());
            sb.setLength(0);
            sb.append("Finished mult ");
            sb.append(size);
            System.out.println(sb.toString());


            // Mult line
            List<Double> results2 = new ArrayList<>();
            for (int iteration = 0; iteration < 3; iteration++) {
                double start = System.nanoTime();
                Main.OnMultLine(m1, m2, res, size);
                double stop = System.nanoTime();
                results2.add((stop - start) / 1000000000);
            }
            sb.setLength(0);
            sb.append("../doc/execution_data/mult_line/java_mult_line_result_");
            sb.append(size);
            sb.append(".csv");
            Main.printToCSV(results2, sb.toString());
            sb.setLength(0);
            sb.append("Finished mult line ");
            sb.append(size);
            System.out.println(sb.toString());
        }
    }

    // Normal Multiplication
    public static void OnMult(int[][] m1, int[][] m2, int[][] res, int n) {

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    res[i][j] += m1[i][k] * m2[k][j];
                }
            }
        }

    }

    // Line Multiplication
    public static void OnMultLine(int[][] m1, int[][] m2, int[][] res, int n) {

        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                for (int j = 0; j < n; j++) {
                    res[i][j] += m1[i][k] * m2[k][j];
                }
            }
        }

    }

    // Block Multiplication
    public static void OnMultBlock(int[][] m1, int[][] m2, int[][] res, int n, int blockSize) {

        if( n % blockSize != 0 || n % blockSize != 0){
            System.out.println("Error");

        }    
        
        
        for(int i0 = 0; i0 < n; i0 += blockSize ){
            for(int i1 = 0; i1 < n; i1 += blockSize){
                for(int i2 = 0; i2 < n; i2 += blockSize){
                    for(int i = i0; i < i0 + blockSize; i++){
                        for(int j = i1; j < i1 + blockSize; j++){
                            for(int k = i2; k < i2 + blockSize; k++){
                                res[i][j] += m1[i][k] * m2[k][j]; 
                            }
                        }
                    }
                }
            }
        }    
       
    }

    public static void main(String[] args) {

        try (Scanner input = new Scanner(System.in)) {
            while (true) {
                double start = 0, stop = 0;
                int option = 0;

                // User input
                while (true) {
                    System.out.println("\n\nChoose method:\n1. Multiplication\n2. Line Multiplication\n3. Block Multiplication\n4. Benchmark mode\n5. Exit");
                    option = input.nextInt();
                    if (option >= 1 && option <= 5) break;
                    System.out.println("\nInvalid option (choose between 1 2 or 3)");
                }

                if (option == 5) break;

                if (option == 4) {
                    benchmark();


                    return;
                }

                System.out.print("Matrix dimensions:");
                int size = input.nextInt();


                // Matrix initialization
                int[][] m1 = new int[size][size];
                int[][] m2 = new int[size][size];
                int[][] res = new int[size][size];
                Main.initializeMatrixes(m1, m2, res, size);

                // Calculations
                switch (option) {
                    case 1: {
                        start = System.nanoTime();
                        Main.OnMult(m1, m2, res, size);
                        stop = System.nanoTime();
                        break;
                    }
                    case 2: {
                        start = System.nanoTime();
                        Main.OnMultLine(m1, m2, res, size);
                        stop = System.nanoTime();
                        break;
                    }
                    case 3: {
                        System.out.print("\nBlock size:");
                        int noBlocks = input.nextInt();
                        start = System.nanoTime();
                        Main.OnMultBlock(m1, m2, res, size, noBlocks);
                        stop = System.nanoTime();
                        break;
                    }
                    case 4: {

                    }
                    default: {
                        System.out.println("\nInternal error: Invalid option");
                        return;
                    }
                }

                // Results
                Main.partiallyPrintMatrix(res, size);

                float time = (float)(stop - start) / 1000000000;
                System.out.println("Time:" + (time) + "seconds");
                
            }
            input.close();
        }
    }
}

File: assign1/src/main.cpp
#include <cstdio>
#include <fstream>
#include <iostream>
#include <ctime>
#include <cmath>
#include <cstdlib>
#include <papi.h>
#include <sstream>
#include "papi_macro.hpp"

using namespace std;

#define SYSTEMTIME clock_t

#define NUMBER_OF_TRIES 3
#define BLOCK_SIZE 512

void printResultMatrix(double* phc, size_t& matrixSize, ostream& debugStream = cerr) {
  for (size_t i = 0; i < 1; i++)
  {
    for (size_t j = 0; j < min((size_t) 10, matrixSize); j++)
      debugStream << phc[j] << ' ';
  }
  debugStream << ';';
}

int getBlockSize() {
  int blockSize;
  
  cout << "Block Size? ";
  cin >> blockSize;

  return blockSize;
}

void resetMatrix(double* pha, double* phb, double* phc, const int& matrixSize) {
  for (size_t i = 0; i < matrixSize; i++) {
    for (size_t j = 0; j < matrixSize; j++) {
      pha[i * matrixSize + j] = (double)1.0;
      phb[i * matrixSize + j] = (double)(i + 1);
      phc[i * matrixSize + j] = (double)0.0;
    }
  }
}

void allocateAndInitMatrix(double** pha, double** phb, double** phc, const int& matrixSize) {
  *pha = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
  *phb = (double *)malloc((matrixSize * matrixSize) * sizeof(double));
  *phc = (double *)malloc((matrixSize * matrixSize) * sizeof(double));

  resetMatrix(*pha, *phb, *phc, matrixSize);
}

void freeAllMatrix(double* pha, double* phb, double* phc) {
  free(pha);
  free(phb);
  free(phc);
}

void OnMult(double* pha, double* phb, double* phc, size_t matrixSize, size_t blockSize = 0, double *timeRes = NULL)
{
  SYSTEMTIME Time1, Time2;

  char st[100];
  int i, j, k;

  Time1 = clock();

  for (i = 0; i < matrixSize; i++)
  {
    for (j = 0; j < matrixSize; j++)
    {
      for (k = 0; k < matrixSize; k++)
      {
        phc[i * matrixSize + j] += pha[i * matrixSize + k] * phb[k * matrixSize + j];
      }
    }
  }

  Time2 = clock();

  if (timeRes != NULL)
  {
    *timeRes = (double)(Time2 - Time1) / CLOCKS_PER_SEC;
  }
}

// add code here for line x line matriz multiplication
void OnMultLine(double* pha, double* phb, double* phc, size_t matrixSize, size_t blockSize = 0, double *timeRes = NULL)
{
  SYSTEMTIME Time1, Time2;

  char st[100];
  int i, j, k;

  Time1 = clock();

  for (i = 0; i < matrixSize; i++)
  {
    for (j = 0; j < matrixSize; j++)
    {
      for (k = 0; k < matrixSize; k++)
      {
        phc[i * matrixSize + k] += pha[i * matrixSize + j] * phb[j * matrixSize + k];
      }
    }
  }

  Time2 = clock();

  if (timeRes != NULL)
  {
    *timeRes = (double)(Time2 - Time1) / CLOCKS_PER_SEC;
  }
}

// add code here for block x block matriz multiplication
void OnMultBlock(double* pha, double* phb, double* phc, size_t matrixSize, size_t blockSize, double *timeRes = NULL)
{
  cout << "Running with block size " << blockSize << endl;

  if (matrixSize % blockSize != 0)
  {
    return;
  }

  SYSTEMTIME Time1, Time2;

  char st[100];
  int i0, j0, k0, i, j, k, numberOfBlocks;

  numberOfBlocks = (matrixSize / blockSize);

  Time1 = clock();

  for (i0 = 0; i0 < matrixSize; i0 += blockSize)
  {
    for (j0 = 0; j0 < matrixSize; j0 += blockSize)
    {
      for (k0 = 0; k0 < matrixSize; k0 += blockSize)
      {
        for (i = i0; i < i0 + blockSize; i++)
        {
          for (j = j0; j < j0 + blockSize; j++)
          {
            for (k = k0; k < k0 + blockSize; k++)
            {
              phc[i * matrixSize + k] += pha[i * matrixSize + j] * phb[j * matrixSize + k];
            }
          }
        }
      }
    }
  }

  Time2 = clock();

  if (timeRes != NULL)
  {
    *timeRes = (double)(Time2 - Time1) / CLOCKS_PER_SEC;
  }
}

void handle_error(int retval)
{
  printf("PAPI error %d: %s\n", retval, PAPI_strerror(retval));
  exit(1);
}

void init_papi(int *EventSet) {
    int retval = PAPI_library_init(PAPI_VER_CURRENT);
    if (retval != PAPI_VER_CURRENT && retval < 0) {
        printf("PAPI library version mismatch!\n");
        exit(1);
    }
    if (retval < 0)
        handle_error(retval);

    std::cout << "PAPI Version Number: MAJOR: " << PAPI_VERSION_MAJOR(retval)
              << " MINOR: " << PAPI_VERSION_MINOR(retval)
              << " REVISION: " << PAPI_VERSION_REVISION(retval) << "\n";

    retval = PAPI_create_eventset(EventSet);
    if (retval != PAPI_OK)
        cout << "ERROR: create eventset" << endl;

    //TODO: Estudar efeitos dos eventos PAPI aqui


    retval = PAPI_add_event(*EventSet, PAPI_L1_DCM);
    if (retval != PAPI_OK)
        cout << "ERROR: PAPI_L1_DCM" << endl;

    retval = PAPI_add_event(*EventSet, PAPI_L2_DCM);
    if (retval != PAPI_OK)
        cout << "ERROR: PAPI_L2_DCM" << endl;

    retval = PAPI_add_event(*EventSet, PAPI_L2_DCA);
    if (retval != PAPI_OK)
        cout << "ERROR: PAPI_L2_DCA" << endl;

    retval = PAPI_add_event(*EventSet, PAPI_L3_DCA);
    if (retval != PAPI_OK)
        cout << "ERROR: PAPI_L3_DCA" << endl;
}

void destroy_papi(int EventSet) {
  int ret;
  ret = PAPI_remove_event(EventSet, PAPI_L1_DCM);
  if (ret != PAPI_OK)
    std::cout << "FAIL remove event" << endl;

  ret = PAPI_remove_event(EventSet, PAPI_L2_DCM);
  if (ret != PAPI_OK)
    std::cout << "FAIL remove event" << endl;

  ret = PAPI_destroy_eventset(&EventSet);
  if (ret != PAPI_OK)
    std::cout << "FAIL destroy" << endl;
}

void benchmark(string filePrefix, size_t initialSize, size_t finalSize, size_t step, size_t blockSize, int EventSet, void (*action)(double*, double*, double*, size_t, size_t, double*))
{
  double* pha;
  double* phb;
  double* phc;

  long long eventValues[NUMBER_OF_PAPI_EVENTS];
  long long executionEventValues[NUMBER_OF_PAPI_EVENTS];
  char filename[256];
  for (size_t matrixSize = initialSize; matrixSize <= finalSize; matrixSize += step)
  {
    allocateAndInitMatrix(&pha, &phb, &phc, matrixSize);
    sprintf(filename, "%s_result_%ld.csv", filePrefix.c_str(), matrixSize);
    stringstream ss;
    ss << "../doc/execution_data/" << filePrefix << "/" << filename;
    ofstream benchmarkFile(ss.str());
    benchmarkFile << "Iteration;Result Matrix;";
    for (size_t i = 0; i < NUMBER_OF_PAPI_EVENTS; i++ ) {
      benchmarkFile << "Event " << i << ';';
    }
    benchmarkFile << "Execution Time" << endl;
    double avgExecutionTime = 0;
    for(size_t i = 0; i < NUMBER_OF_PAPI_EVENTS; i++)
    {
      eventValues[i] = 0;
    }
    for (size_t exe = 0; exe < NUMBER_OF_TRIES; exe++) {
      benchmarkFile << exe << ';';
      double executionTime = 0;
      start_papi_event_counter(EventSet);
      action(pha, phb, phc, matrixSize, blockSize, &executionTime);
      stop_papi_event_counter(EventSet, executionEventValues);
      printResultMatrix(phc, matrixSize, benchmarkFile);
      for (size_t i = 0; i < NUMBER_OF_PAPI_EVENTS; i++) {
        benchmarkFile << executionEventValues[i] << ';';
        eventValues[i] += executionEventValues[i];
      }
      benchmarkFile << executionTime << std::endl;
      avgExecutionTime += executionTime;
      resetMatrix(pha, phb, phc, matrixSize);
      reset_papi_event_counter(EventSet);
    }
    freeAllMatrix(pha, phb, phc);
    avgExecutionTime /= NUMBER_OF_TRIES;
    for (int i = 0; i < NUMBER_OF_PAPI_EVENTS; i++) {
      eventValues[i] = (long long) round(eventValues[i] / (double) NUMBER_OF_TRIES);
    }
    benchmarkFile << "Average;;";
    for (size_t i = 0; i < NUMBER_OF_PAPI_EVENTS; i++) {
      benchmarkFile << eventValues[i] << ";";
    }
    benchmarkFile << avgExecutionTime << std::endl;
    benchmarkFile.close();
  }
}

int main(int argc, char *argv[])
{
  double* pha;
  double* phb;
  double* phc;

  pha = phb = phc =  NULL;

  char c;
  size_t lin, col, blockSize;
  int op;

  int EventSet = PAPI_NULL;
  long long values[NUMBER_OF_PAPI_EVENTS];
  
  init_papi(&EventSet);

  op = 1;
  do
  {
    cout << endl
         << "1. Multiplication" << endl;
    cout << "2. Line Multiplication" << endl;
    cout << "3. Block Multiplication" << endl;
    cout << "4. Benchmark Multiplication(600 to 3000 step 400)" << endl;
    cout << "5. Benchmark Line Multiplication(600 to 3000 step 400)" << endl;
    cout << "6. Benchmark Line Multiplication(4096 to 10240 step 2048)" << endl;
    cout << "7. Benchmark Block Multiplication(4096 to 10240 step 2048)" << endl;
    cout << "Selection?: ";
    cin >> op;
    if (op == 0)
      break;
    if (op < 4) {
      printf("Dimensions: lins=cols ? ");
      cin >> lin;
      col = lin;
    }

    size_t blockSize;

    if (op < 4) {
      allocateAndInitMatrix(&pha, &phb, &phc, lin);
    }
    double timeRes;
    switch (op)
    {
      case 1:
        OnMult(pha, phb, phc, lin, 0, &timeRes);
        std::cout << "Execution time: " << timeRes << "seconds" << std::endl;
        break;
      case 2:
        OnMultLine(pha, phb, phc, lin, 0, &timeRes);
        std::cout << "Execution time: " << timeRes << "seconds" << std::endl;
        break;
      case 3:
        blockSize = getBlockSize();
        if (blockSize != -1) {
          OnMultBlock(pha, phb, phc, lin, blockSize, &timeRes);
        } else {
          cout << "Invalid block size!" << endl;
        }
        std::cout << "Execution time: " << timeRes << "seconds" << std::endl;
        break;
      case 4:
        benchmark("mult", 600, 3000, 400, blockSize, EventSet, OnMult);
        break;
      case 5:
        benchmark("mult_line", 600, 3000, 400, blockSize, EventSet, OnMultLine);
        break;
      case 6:
        benchmark("mult_line", 4096, 10240, 2048, blockSize, EventSet, OnMultLine);
        break;
      case 7: {
        blockSize = getBlockSize();
        if (blockSize != -1) {
          char multBlockFilePrefix[50];
          sprintf(multBlockFilePrefix, "mult_block_%d", blockSize);
          benchmark(multBlockFilePrefix, 4096, 10240, 2048, blockSize, EventSet, OnMultBlock);
        } else {
          cout << "Invalid block size!" << endl;
        }
        break;
      }
    }

    if (op < 4) {
      freeAllMatrix(pha, phb, phc);
    }

  } while (op != 0);

  destroy_papi(EventSet);
}